### DISCUSSION TITLE
How can we handle DB Transactions?

### DISCUSSION THREAD

**Discussion by yasalkhatib** [lucidarch/lucid#4] [Q&A] [answered]
I have a question regarding DB Transactions.

### Example
Money Transfer from account A to Account B.
Feature: TranserMoneyFeature
Jobs: 
     1. Withdraw Money from Account A
     2. Deposit money in Account B

What I would want to achieve is to make the two jobs above as a unit. All jobs or none. 

Would you please suggest a way to do that?

**Comment by Mulkave** [comment, marked as answer]
If i understand correctly you're looking into a reversal in case one of these jobs fail? If so, this is an interesting case!

This makes a good case for an [operation](https://lucid-architecture.gitbook.io/docs/components#658b4769-cef3-4c8f-ac18-d4a582dee61e).

**Important**
Before going into the details, I'd like to highlight that these are usually handled at the DB layer and not at the architecture layer, where you would create a DB transaction which will revert itself when things don't go as planned.

---

However, if you were to do it manually at the architecture level, here's one suggestion to make it happen:

- `TransferMoneyOperation($fromAccount, $toAccount)` -> returns `TransferStatus` object summarising the jobs' results
   - `WithdrawMoneyFromAccountJob($fromAccount, $amount)`
   - `DepositMoneyInAccountJob($toAccount, $amount)`

Based on the result from the `TransferStatus` which could hold as many details as required about the failure if it were to happen, the next operation can take place.

**If failed** run `RevertMoneyTransferOperation($transferStatus, $fromAccount, $toAccount, $amount)` which contains two jobs that run conditionally.
   - If withdrawal error: `RevertMoneyWithdrawalFromAccountJob($fromAccount, $amount)`
   - If deposit error: `HandleDepositErrorJob($fromAccount, $toAccount, $amount)`

The reason why the second job isn't named with the same convention (revert etc.) is because it contains more work than 

---

**General Suggestions**

- **Build rich Jobs**: Jobs don't need to be shallow (performing single-line jobs or contain simple code) but rather should be deep/rich in order to provide reusability
- **Use objects as an exchange structure**: Exchanging objects between components ensures the robustness of the application. For example: 
```
TransferMoneyOperation(Account $fromAccount, Account $toAccount)
WithdrawMoneyFromAccountJob(Account $fromAccount, [float|int] $amount)
```
---

Please let me know your thoughts on the above @yasalkhatib
